### PR TITLE
Client side remote credential guard

### PR DESCRIPTION
**Pull request by hardening**
This PR implements the client-side part of the remote credential guard feature as described in MS-RDPEAR. The 2 main changes are: 

* shipping the TSRemoteGuardCreds in NLA, 
* and implement the rdpear channel that allows LSASS to remote all the calls to our client. 

For now it's UNIX only as the windows implementation would be implemented in a completely different way.

To test, you may establish you ccache and then connect with (RCG enabled on the server):
`xfreerdp /remoteGuard /u:<user> /d:<domain> /v<server>`

That should log you in, and in the session you should not be asked for credentials when doing `mstsc /remoteGuard /v:<other server>`.

### CODE REVIEW

File: winpr/libwinpr/sspi/Kerberos/kerberos.c [hunk 1]
```diff
@@ -216,6 +223,20 @@ static BOOL isValidIP(const char* ipAddress)
 	return isValidIPv4(ipAddress) || isValidIPv6(ipAddress);
 }
 
+static int build_krbtgt(krb5_context ctx, krb5_data* realm, krb5_principal target)
+{
+	/* "krbtgt/" + realm + "@" + realm */
+	int len = 7 + realm->length + 1 + realm->length + 1;
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/Kerberos/kerberos.c:229]: should be `size_t`, constants suffixed by `ULL` to force the data type to be 64bit

File: winpr/libwinpr/sspi/Kerberos/kerberos.c [hunk 2]
```diff
@@ -358,14 +379,32 @@ static SECURITY_STATUS SEC_ENTRY kerberos_AcquireCredentialsHandleA(
 	/* Get initial credentials if required */
 	if (fCredentialUse & SECPKG_CRED_OUTBOUND)
 	{
-		if (krb_log_exec(krb5glue_get_init_creds, ctx, principal, ccache, krb5_prompter, password,
-		                 krb_settings))
+		krb5_creds creds;
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/Kerberos/kerberos.c:382]: init with `= { 0 };`

File: winpr/libwinpr/sspi/Kerberos/kerberos.c [hunk 3]
```diff
@@ -442,34 +482,39 @@ static SECURITY_STATUS SEC_ENTRY kerberos_AcquireCredentialsHandleW(
...
-
-	credentials = sspi_SecureHandleGetLowerPointer(phCredential);
-	if (!credentials)
-		return SEC_E_INVALID_HANDLE;
-
-	if (krb5_init_context(&ctx))
-		return SEC_E_INTERNAL_ERROR;
+static void credentials_unref(KRB_CREDENTIALS* credentials)
+{
+	if (InterlockedDecrement(&credentials->refCount))
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/Kerberos/kerberos.c:488]: maybe do a `WINPR_ASSERT(credentials)` or just ignore `NULL` arguments?

File: winpr/libwinpr/sspi/Kerberos/kerberos.c [hunk 4]
```diff
@@ -442,34 +482,39 @@ static SECURITY_STATUS SEC_ENTRY kerberos_AcquireCredentialsHandleW(
...
 
+	krb5_free_context(credentials->ctx);
 	free(credentials);
+}
+#endif
+
+static SECURITY_STATUS SEC_ENTRY kerberos_FreeCredentialsHandle(PCredHandle phCredential)
+{
+#ifdef WITH_KRB5
+	KRB_CREDENTIALS* credentials;
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/Kerberos/kerberos.c:511]: merge with initialization line below

File: winpr/libwinpr/sspi/Kerberos/kerberos.c [hunk 5]
```diff
@@ -753,9 +801,8 @@ static SECURITY_STATUS SEC_ENTRY kerberos_InitializeSecurityContextA(
     PCtxtHandle phNewContext, PSecBufferDesc pOutput, ULONG* pfContextAttr, PTimeStamp ptsExpiry)
 {
 #ifdef WITH_KRB5
-	KRB_CREDENTIALS* credentials = NULL;
-	KRB_CONTEXT* context = NULL;
-	KRB_CONTEXT new_context = { 0 };
+	KRB_CREDENTIALS* credentials;
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/Kerberos/kerberos.c:804]: initialize with `=NULL;`

File: winpr/libwinpr/sspi/Kerberos/kerberos.c [hunk 6]
```diff
@@ -753,9 +801,8 @@ static SECURITY_STATUS SEC_ENTRY kerberos_InitializeSecurityContextA(
...
+	KRB_CONTEXT* context;
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/Kerberos/kerberos.c:805]: initialize with `=NULL;`

File: winpr/libwinpr/sspi/Kerberos/kerberos.c [hunk 7]
```diff
@@ -1336,6 +1378,18 @@ static KRB_CONTEXT* get_context(PCtxtHandle phContext)
 	return sspi_SecureHandleGetLowerPointer(phContext);
 }
 
+static BOOL copy_krb5_data(krb5_data* data, PUCHAR* ptr, ULONG* psize)
+{
+	*ptr = (PUCHAR)malloc(data->length);
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/Kerberos/kerberos.c:1383]: `WINPR_ASSERT(data);`
`WINPR_ASSER(ptr);`
`WINPR_ASSERT(psize);`

File: winpr/libwinpr/sspi/Kerberos/kerberos.c [hunk 8]
```diff
@@ -1351,72 +1405,177 @@ static SECURITY_STATUS SEC_ENTRY kerberos_DeleteSecurityContext(PCtxtHandle phCo
...
+                                                  KRB_CREDENTIALS* credentials,
+                                                  KERB_TICKET_LOGON* ticketLogon)
+{
+	krb5_creds matchCred = { 0 };
+	krb5_auth_context authContext = NULL;
+	int getCredsFlags = KRB5_GC_CACHED;
+	BOOL firstRun = TRUE;
+	krb5_creds* hostCred = NULL;
+	SECURITY_STATUS ret = SEC_E_INSUFFICIENT_MEMORY;
+	int rv;
```
Comment (akallabeth) [inline winpr/libwinpr/sspi/Kerberos/kerberos.c:1487]: merge with initialization below

File: libfreerdp/core/nla.c [hunk 9]
```diff
@@ -1405,6 +1405,68 @@ static BOOL nla_read_ts_credentials(rdpNla* nla, SecBuffer* data)
 	return ret;
 }
 
+static BOOL nla_write_KERB_TICKET_LOGON(wStream* s, const KERB_TICKET_LOGON* ticket)
+{
+	if (!Stream_EnsureRemainingCapacity(s, (4 * 4) + 16 + ticket->ServiceTicketLength +
```
Comment (akallabeth) [inline libfreerdp/core/nla.c:1410]: `WINPR_ASSERT(ticket);`
suffix constants with `ULL` to force 64bit

File: libfreerdp/core/nla.c [hunk 10]
```diff
@@ -1405,6 +1405,68 @@ static BOOL nla_read_ts_credentials(rdpNla* nla, SecBuffer* data)
...
+	Stream_Write_UINT64(s, 0x20 + ticket->ServiceTicketLength); /* offset of TGT in packet */
+
+	Stream_Write(s, ticket->ServiceTicket, ticket->ServiceTicketLength);
+	Stream_Write(s, ticket->TicketGrantingTicket, ticket->TicketGrantingTicketLength);
+	return TRUE;
+}
+
+BOOL nla_get_KERB_TICKET_LOGON(rdpNla* nla, KERB_TICKET_LOGON* logonTicket)
+{
+	SecurityFunctionTable* table;
```
Comment (akallabeth) [inline libfreerdp/core/nla.c:1429]: Initialize variables
`WINPR_ASSERT(nla);`

File: libfreerdp/core/nla.c [hunk 11]
```diff
@@ -2107,3 +2192,30 @@ UINT32 nla_get_sspi_error(rdpNla* nla)
...
+	WINPR_ASSERT(inBuffer);
+	WINPR_ASSERT(outBuffer);
+	return credssp_auth_decrypt(nla->auth, inBuffer, outBuffer, nla->recvSeqNum++);
+}
+
+SECURITY_STATUS nla_setKerbAttribute(rdpNla* nla, DWORD ulAttr, PVOID pBuffer)
+{
+	WINPR_ASSERT(nla);
+
+	SecurityFunctionTable* table;
```
Comment (akallabeth) [inline libfreerdp/core/nla.c:2216]: initialzie variables

File: client/common/cmdline.c [hunk 12]
```diff
@@ -4399,6 +4400,12 @@ static int freerdp_client_settings_parse_command_line_arguments_int(
 			if (!freerdp_settings_set_bool(settings, FreeRDP_RestrictedAdminModeRequired, enable))
 				return fail_at(arg, COMMAND_LINE_ERROR);
 		}
+		CommandLineSwitchCase(arg, "remoteGuard")
+		{
+			if (!freerdp_settings_set_bool(settings, FreeRDP_RemoteCredentialGuard, TRUE) ||
+			    !freerdp_settings_set_bool(settings, FreeRDP_ExtSecurity, TRUE))
+				return COMMAND_LINE_ERROR;
```
Comment (akallabeth) [inline client/common/cmdline.c:4407]: use `fail_at(arg, COMMAND_LINE_ERROR)` to give useful feedback about where parsing failed

File: channels/urbdrc/client/libusb/libusb_udevice.c [hunk 13]
```diff
@@ -290,11 +290,7 @@ static void LIBUSB_CALL func_iso_callback(struct libusb_transfer* transfer)
 			/* fallthrough */
 			WINPR_FALLTHROUGH
 		case LIBUSB_TRANSFER_CANCELLED:
-			/* fallthrough */
```
Comment (akallabeth) [inline channels/urbdrc/client/libusb/libusb_udevice.c:293]: accidental change?

File: channels/urbdrc/client/libusb/libusb_udevice.c [hunk 14]
```diff
@@ -290,11 +290,7 @@ static void LIBUSB_CALL func_iso_callback(struct libusb_transfer* transfer)
...
-			WINPR_FALLTHROUGH
 		case LIBUSB_TRANSFER_TIMED_OUT:
-			/* fallthrough */
-			WINPR_FALLTHROUGH
```
Comment (akallabeth) [inline channels/urbdrc/client/libusb/libusb_udevice.c:297]: accidental change?

File: channels/rdpear/client/rdpear_main.c [hunk 15]
```diff
@@ -0,0 +1,944 @@
...
+#include <freerdp/freerdp.h>
+#include <freerdp/addin.h>
+#include <freerdp/client/channels.h>
+#include <freerdp/channels/log.h>
+#include <freerdp/channels/rdpear.h>
+
+#define TAG CHANNELS_TAG("rdpear.client")
+
+/* defined in libkrb5 */
+krb5_error_code encode_krb5_authenticator(const krb5_authenticator* rep, krb5_data** code_out);
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:45]: `extern`?

File: channels/rdpear/client/rdpear_main.c [hunk 16]
```diff
@@ -0,0 +1,944 @@
...
+krb5_error_code encode_krb5_ap_rep(const krb5_ap_rep* rep, krb5_data** code_out);
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:46]: `extern`?

File: channels/rdpear/client/rdpear_main.c [hunk 17]
```diff
@@ -0,0 +1,944 @@
...
+	BOOL ret = FALSE;
+	wStream* unencodedContent = rdpear_encodePayload(packageType, payload);
+	if (!unencodedContent)
+		goto out;
+
+	size_t unencodedLen = Stream_GetPosition(unencodedContent);
+	SecBuffer inBuffer = { unencodedLen, SECBUFFER_DATA, Stream_Buffer(unencodedContent) };
+	SecBuffer cryptedBuffer = { 0 };
+
+	if (!freerdp_nla_encrypt(rdpear->rdp_context, &inBuffer, &cryptedBuffer))
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:191]: `WINPR_ASSERT` before dereference

File: channels/rdpear/client/rdpear_main.c [hunk 18]
```diff
@@ -0,0 +1,944 @@
...
+		goto out;
+
+	wStream* finalStream = Stream_New(NULL, 200);
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:194]: `NULL` check?

File: channels/rdpear/client/rdpear_main.c [hunk 19]
```diff
@@ -0,0 +1,944 @@
...
+		goto out;
+	// ComputeTgsChecksumReq_dump(WLog_Get(""), WLOG_DEBUG, &req);
+
+	krb5_error_code rv =
+	    kerb_do_checksum(rdpear->krbContext, req.Key, KRB5_KEYUSAGE_TGS_REQ_AUTH_CKSUM,
+	                     req.ChecksumType, req.requestBody, &checksum);
+	if (rv)
+		goto out;
+
+	wStream* asn1Payload = rdpear_enc_Checksum(req.ChecksumType, &checksum);
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:292]: declare `wStream* asn1Payload` before first `goto`!

File: channels/rdpear/client/rdpear_main.c [hunk 20]
```diff
@@ -0,0 +1,944 @@
...
+out:
+	krb5_free_data_contents(rdpear->krbContext, &encrypted);
+	ndr_destroy_BuildEncryptedAuthDataReq(rcontext, NULL, &req);
+	Stream_Free(asn1Payload, FALSE);
+	return TRUE;
+}
+
+static char* KERB_RPC_UNICODESTR_to_charptr(const RPC_UNICODE_STRING* src)
+{
+	return ConvertWCharNToUtf8Alloc(src->Buffer, src->strLength, NULL);
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:349]: `WINPR_ASSERT`

File: channels/rdpear/client/rdpear_main.c [hunk 21]
```diff
@@ -0,0 +1,944 @@
...
+}
+
+static BOOL extractAuthData(const KERB_ASN1_DATA* src, krb5_authdata* authData, BOOL* haveData)
+{
+	WinPrAsn1Decoder dec, dec2, dec3;
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:354]: Initialize variables, assert arguments

File: channels/rdpear/client/rdpear_main.c [hunk 22]
```diff
@@ -0,0 +1,944 @@
...
+	authData->ad_type = adType;
+	authData->length = os.len;
+	authData->contents = os.data;
+	*haveData = TRUE;
+	return TRUE;
+}
+
+static BOOL extractChecksum(const KERB_ASN1_DATA* src, krb5_checksum* dst)
+{
+	WinPrAsn1Decoder dec, dec2;
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:386]: Initialize varialbes, assert arguments

File: channels/rdpear/client/rdpear_main.c [hunk 23]
```diff
@@ -0,0 +1,944 @@
...
+	dst->length = os.len;
+	dst->contents = os.data;
+	return TRUE;
+}
+
+#define FILETIME_TO_UNIX_OFFSET_S 11644473600UL
+
+UINT64 krb5_time_to_FILETIME(const krb5_timestamp* ts, krb5_int32 usec)
+{
+	return ((*ts + FILETIME_TO_UNIX_OFFSET_S) * (1000 * 1000) + usec) * 10;
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:409]: assert arguments

File: channels/rdpear/client/rdpear_main.c [hunk 24]
```diff
@@ -0,0 +1,944 @@
...
+	krb5_principal_data client = { 0 };
+
+	*pstatus = ERROR_INVALID_DATA;
+	WLog_DBG(TAG, "-> CreateApReqAuthenticator");
+
+	if (!ndr_read_CreateApReqAuthenticatorReq(rcontext, s, NULL, &req) ||
+	    !ndr_treat_deferred_read(rcontext, s))
+		goto out;
+
+	krb5_authdata authdata;
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:439]: Initialize variables

File: channels/rdpear/client/rdpear_main.c [hunk 25]
```diff
@@ -0,0 +1,944 @@
...
+	krb5_authdata* authDataPtr[2] = { &authdata, NULL };
+	BOOL haveData;
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:441]: Initialize variables

File: channels/rdpear/client/rdpear_main.c [hunk 26]
```diff
@@ -0,0 +1,944 @@
...
+	for (int i = 0; i < client.length; i++)
+	{
+		client.data[i].data = KERB_RPC_UNICODESTR_to_charptr(&req.ClientName->Names[i]);
+		client.data[i].length = strlen(client.data[i].data);
+	}
+	client.realm.data = KERB_RPC_UNICODESTR_to_charptr(req.ClientRealm);
+	client.realm.length = strlen(client.realm.data);
+	authent.client = &client;
+
+	krb5_checksum checksum;
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:477]: Initialize variables

File: channels/rdpear/client/rdpear_main.c [hunk 27]
```diff
@@ -0,0 +1,944 @@
...
+	if (subkey)
+		krb5_free_keyblock(rdpear->krbContext, subkey);
+	ndr_destroy_CreateApReqAuthenticatorReq(rcontext, NULL, &req);
+	Stream_Free(asn1EncodedAuth, FALSE);
+	return TRUE;
+}
+
+static BOOL rdpear_findEncryptedData(const KERB_ASN1_DATA* src, int* penctype, krb5_data* data)
+{
+	WinPrAsn1Decoder dec, dec2;
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:543]: Initialize variables, assert arguments

File: channels/rdpear/client/rdpear_main.c [hunk 28]
```diff
@@ -0,0 +1,944 @@
...
+		WLog_ERR(TAG, "StrengthenKey not supported yet");
+		goto out;
+	}
+
+	WLog_DBG(TAG, "-> UnpackKdcReplyBody: KeyUsage=0x%x PDU=0x%x", req.KeyUsage, req.Pdu);
+	// WLog_DBG(TAG, "encryptedPayload=");
+	// winpr_HexDump(TAG, WLOG_DEBUG, req.EncryptedData->Asn1Buffer,
+	// req.EncryptedData->Asn1BufferHints.count);
+
+	krb5_data asn1Data;
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:584]: Initialize variables

File: channels/rdpear/client/rdpear_main.c [hunk 29]
```diff
@@ -0,0 +1,944 @@
...
+	int encType;
+	if (!rdpear_findEncryptedData(req.EncryptedData, &encType, &asn1Data))
+		goto out;
+
+	resp->KerbProtocolError =
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:589]: assert before dereferenes

File: channels/rdpear/client/rdpear_main.c [hunk 30]
```diff
@@ -0,0 +1,944 @@
...
+	*pstatus = ERROR_INVALID_DATA;
+	if (!ndr_read_DecryptApReplyReq(rcontext, s, NULL, &req) ||
+	    !ndr_treat_deferred_read(rcontext, s))
+		goto out;
+
+	WLog_DBG(TAG, "-> DecryptApReply");
+	// winpr_HexDump(TAG, WLOG_DEBUG, req.EncryptedReply->Asn1Buffer,
+	// req.EncryptedReply->Asn1BufferHints.count);
+
+	krb5_data asn1Data;
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:616]: Initialize variables

File: channels/rdpear/client/rdpear_main.c [hunk 31]
```diff
@@ -0,0 +1,944 @@
...
+	int encType;
+	if (!rdpear_findEncryptedData(req.EncryptedReply, &encType, &asn1Data))
+		goto out;
+
+	resp->Pdu = 0x31;
+	krb5_error_code rv =
+	    kerb_do_decrypt(rdpear->krbContext, req.Key, KRB5_KEYUSAGE_AP_REP_ENCPART, &asn1Data, resp);
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:623]: assert before dereference

File: channels/rdpear/client/rdpear_main.c [hunk 32]
```diff
@@ -0,0 +1,944 @@
...
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpear_on_open(IWTSVirtualChannelCallback* pChannelCallback)
+{
+
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:897]: `WINPR_UNUSED` if on purpose?

File: channels/rdpear/client/rdpear_main.c [hunk 33]
```diff
@@ -0,0 +1,944 @@
...
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT rdpear_on_close(IWTSVirtualChannelCallback* pChannelCallback)
+{
+	UINT error = CHANNEL_RC_OK;
```
Comment (akallabeth) [inline channels/rdpear/client/rdpear_main.c:908]: `WINPR_UNUSED` if on purpose?

File: channels/rdpear/test/TestNdrEar.c [hunk 34]
```diff
@@ -0,0 +1,344 @@
...
+
+#include "../ndr.h"
+#include "../rdpear_common.h"
+
+BYTE *parseHexBlock(const char *str, size_t *plen)
+{
+	WINPR_ASSERT(str);
+	WINPR_ASSERT(plen);
+
+	BYTE *ret = malloc(strlen(str) / 2);
```
Comment (akallabeth) [inline channels/rdpear/test/TestNdrEar.c:11]: `NULL` check

File: channels/rdpear/test/TestNdrEar.c [hunk 35]
```diff
@@ -0,0 +1,344 @@
...
+int TestNdrEarRead(int argc, char* argv[])
+{
+	int retCode = -2;
+
+	/* ====================================================================== */
+	NdrContext* context = ndr_context_new(FALSE, 1);
+	if (!context)
+		return -1;
+
+	wStream staticS;
```
Comment (akallabeth) [inline channels/rdpear/test/TestNdrEar.c:125]: initialize

File: channels/rdpear/test/TestNdr.c [hunk 36]
```diff
@@ -0,0 +1,39 @@
...
+	if (!context)
+		return -1;
+
+	BYTE payload[] = {
+		// == conformant array ==
+		0x02, 0x00, 0x00, 0x00, // (nitems)
+		0x30, 0x00,             // content
+		0x00, 0x00              // (padding)
+	};
+	wStream staticS;
```
Comment (akallabeth) [inline channels/rdpear/test/TestNdr.c:16]: initialize

File: channels/rdpear/test/CMakeLists.txt [hunk 37]
```diff
@@ -0,0 +1,42 @@
+
+set(MODULE_NAME "TestRdpear")
+set(MODULE_PREFIX "TEST_RDPEAR")
+
+set(TEST_RDPEAR_DRIVER TestRdpear.c)
+
+set(TEST_RDPEAR_COMMON_SRCS
+    ../ndr.c
```
Comment (akallabeth) [inline channels/rdpear/test/CMakeLists.txt:8]: maybe create an object library from these as they are used in both, test and client?

File: channels/rdpear/ndr.h [hunk 38]
```diff
@@ -0,0 +1,182 @@
...
+#ifndef CHANNELS_RDPEAR_NDR_H_
+#define CHANNELS_RDPEAR_NDR_H_
+
+#include <winpr/stream.h>
+#include <freerdp/api.h>
+
+typedef struct NdrContext_s NdrContext;
+
+typedef UINT32 ndr_refid;
+#define NDR_PTR_NULL ((UINT32)0x0000)
```
Comment (akallabeth) [inline channels/rdpear/ndr.h:29]: `0UL` is what you are looking for (no cast required)

File: channels/rdpear/rdpear_common.c [hunk 39]
```diff
@@ -0,0 +1,481 @@
...
+#include <stddef.h>
+#include <winpr/print.h>
+#include <freerdp/channels/log.h>
+
+#define TAG CHANNELS_TAG("rdpear")
+
+const char kerberosPackageName[] = {
+	'K', 0, 'e', 0, 'r', 0, 'b', 0, 'e', 0, 'r', 0, 'o', 0, 's', 0
+};
+const char ntlmPackageName[] = { 'N', 0, 'T', 0, 'L', 0, 'M', 0 };
```
Comment (akallabeth) [inline channels/rdpear/rdpear_common.c:32]: sure you don´t need `\0` termination here?

File: channels/rdpear/rdpear_common.c [hunk 40]
```diff
@@ -0,0 +1,481 @@
...
+#include "ndr.h"
+
+#include <stddef.h>
+#include <winpr/print.h>
+#include <freerdp/channels/log.h>
+
+#define TAG CHANNELS_TAG("rdpear")
+
+const char kerberosPackageName[] = {
+	'K', 0, 'e', 0, 'r', 0, 'b', 0, 'e', 0, 'r', 0, 'o', 0, 's', 0
```
Comment (akallabeth) [inline channels/rdpear/rdpear_common.c:30]: sure you don´t need `\0` termination here?

File: channels/rdpear/rdpear_common.c [hunk 41]
```diff
@@ -0,0 +1,481 @@
...
+};
+const char ntlmPackageName[] = { 'N', 0, 'T', 0, 'L', 0, 'M', 0 };
+
+RdpEarPackageType rdpear_packageType_from_name(WinPrAsn1_OctetString* package)
+{
```
Comment (akallabeth) [inline channels/rdpear/rdpear_common.c:35]: assert arguments

**Review by akallabeth — COMMENTED** [review]
Looking good overall.

some cleanups remain (`WINPR_ASSERT`, `WINPR_UNUSED` and variable initialization along with a few conditionally uninitialized variables)

File: channels/urbdrc/client/libusb/libusb_udevice.c (hunk 14 above)
Comment (hardening) [reply channels/urbdrc/client/libusb/libusb_udevice.c:297]: Nope it's a gcc warning I have from a long time (a real fall through is when you have code in the case statement).

File: channels/rdpear/client/rdpear_main.c (hunk 16 above)
Comment (hardening) [reply channels/rdpear/client/rdpear_main.c:46]: These two ones are exported in libkrb5 library although they're not in the `krb5.h` file, I don't think that needs the `extern` modifier:  if they were cited in `krb5.h` file they would not have `extern` ?

File: channels/rdpear/client/rdpear_main.c (hunk 21 above)
Comment (hardening) [reply channels/rdpear/client/rdpear_main.c:354]: For asserts, this is a static functions (so controlled callers). And init would be useless as no cleanup is done at the end (inits are done during function execution with no possible use before init)

File: channels/rdpear/client/rdpear_main.c (hunk 24 above)
Comment (hardening) [reply channels/rdpear/client/rdpear_main.c:439]: init is done by `extractAuthData`

File: channels/rdpear/client/rdpear_main.c (hunk 26 above)
Comment (hardening) [reply channels/rdpear/client/rdpear_main.c:477]: Not needed, the checksum is optional , when set it's an alias to exiting data, and no cleanup is associated with it. So no place for use before init.

File: channels/rdpear/client/rdpear_main.c (hunk 27 above)
Comment (hardening) [reply channels/rdpear/client/rdpear_main.c:543]: Not needed: asserts -> controlled callers, inits -> done during execution and no cleanups

File: channels/rdpear/client/rdpear_main.c (hunk 28 above)
Comment (hardening) [reply channels/rdpear/client/rdpear_main.c:584]: Done two lines below by `rdpear_findEncryptedData`

File: channels/rdpear/client/rdpear_main.c (hunk 29 above)
Comment (hardening) [reply channels/rdpear/client/rdpear_main.c:589]: No, this is a static function with a controlled caller.

**Comment by akallabeth** [comment]
@hardening created a cleaned up version of this branch at https://github.com/akallabeth/FreeRDP/tree/rdpear-cleanup
feel free to cherry-pick

File: channels/rdpear/rdpear_common.c (hunk 39 above)
Comment (hardening) [reply channels/rdpear/rdpear_common.c:32]: Nope this is ASN1 unicode strings

File: channels/rdpear/rdpear_common.c (hunk 40 above)
Comment (hardening) [reply channels/rdpear/rdpear_common.c:30]: Nope this is ASN1 unicode strings

File: channels/urbdrc/client/libusb/libusb_udevice.c (hunk 14 above)
Comment (akallabeth) [reply channels/urbdrc/client/libusb/libusb_udevice.c:297]: then please fix the macro next time ;)
already done with https://github.com/FreeRDP/FreeRDP/pull/10541 so this commit should be dropped.

**Comment by hardening** [comment]
@akallabeth it's more https://learn.microsoft.com/en-us/windows/security/identity-protection/remote-credential-guard?tabs=reg

**Comment by akallabeth** [comment]
@hardening ok, still failing for me (but looks like a configuration issue)
form the pull itself it looks good. (the code cleanup can happen later)

any reason this is still a draft or is this ready from your side?

**Comment by hardening** [comment]
@akallabeth there's still a few memory leaks, that I'm looking at

**Comment by akallabeth** [comment]
@hardening always failing at 
1. `krb5_get_credentials (Matching credential not found [-1765328243])`
2. `krb5_fwd_tgt_creds` with `(KDC can't fulfill requested option [-1765328371])`

**Comment by fifthdegree** [comment]
> @hardening always failing at 
> 1. `krb5_get_credentials (Matching credential not found [-1765328243])`
> 2. `krb5_fwd_tgt_creds` with `(KDC can't fulfill requested option [-1765328371])`
> 
> I have configured the remote servers the way that is suggested in your link (the two GPO settings)

The first is expected unless you happen to already have credentials to the server in your cache. @hardening perhaps you should change the error logging to make this clear.

**Comment by hardening** [comment]
@fifthdegree we have dealt with that off github
I'll close that PR and start a new one as there's way too many comments (and hopefully the PR will not have too many more changes)

**Pull request closed without merging**